### PR TITLE
fix(agent): project public context graph metadata

### PR DESCRIPTION
**Pull request by lupuszr**
Focused replay of the changes introduced by #1866 onto the current `main` branch.

## What changed

- write the canonical public Context Graph proof into the root `_meta` graph during create/register flows
- repair legacy creator-owned public graph metadata during startup
- keep public/private admission fail-closed when metadata is missing or conflicting
- add embedded Oxigraph and SPARQL HTTP regression coverage

## Scope

This branch contains one commit and the same 11-file functional scope as #1866. `main` did not yet contain `context-graph-public-meta-proof.ts`, so that module is added while resolving the original commit onto `main`; no `testnet-canary` history is included.

## Validation

- `pnpm build`
- 104 focused agent tests across metadata repair, discovery, registration, and issue #865 coverage

### CODE REVIEW

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 1]
```diff
@@ -142,6 +142,8 @@ import {
 } from '@origintrail-official/dkg-query';
 import { DKGAgentWallet, type AgentWallet } from './agent-wallet.js';
 import { buildAuthoritativePrivateMetaAskQuery } from './context-graph-private-meta-proof.js';
+import { buildAuthoritativePublicMetaAskQuery } from './context-graph-public-meta-proof.js';
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts:145]: 🔴 Bug: Public `_meta` proof is written but not enforced by admission

**What's wrong**
The PR adds and backfills canonical public proof quads, but production admission does not consult that proof. As a result, stale or ontology-only public declarations can still be treated as confirmed metadata, bypassing the new root `_meta` authority model the change is trying to establish.

**Example**
A subscribed graph with `metaSynced: false` and only an ONTOLOGY declaration `<did:dkg:context-graph:x> rdf:type dkg:ContextGraph` still passes `hasConfirmedMetaState()`, so `refreshMetaSyncedFlags()` or the gossip race-opener can mark it synced/admit SWM even though `<did:dkg:context-graph:x/_meta>` lacks the new canonical public proof.

**Suggested direction**
Use the new public proof query in `hasConfirmedMetaState` before treating a non-private graph as confirmed, and narrow the ontology-only fallback to the explicit cases that are still meant to bypass root `_meta` proof.

**For Agents**
Look at `packages/agent/src/dkg-agent-lifecycle.ts::hasConfirmedMetaState`. Wire `buildAuthoritativePublicMetaAskQuery(contextGraphId)` into the public confirmation path, preserve any intentionally local/system bootstrap exceptions, and prove that ontology-only public declarations no longer set `metaSynced` or open SWM admission while `_meta` public proof does.

File: packages/agent/src/dkg-agent-context-graph.ts [hunk 2]
```diff
@@ -553,6 +554,9 @@ export class ContextGraphMethods extends DKGAgentBase {
       { subject: contextGraphUri, predicate: DKG_ONTOLOGY.DKG_REGISTRATION_STATUS, object: `"unregistered"`, graph: cgMetaGraph },
       { subject: contextGraphUri, predicate: DKG_ONTOLOGY.DKG_CURATOR, object: curatorDid, graph: cgMetaGraph },
     );
+    if (!isCurated && !opts.private) {
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-context-graph.ts:557]: 🟡 Issue: Centralize public context-graph definition writes instead of patching each writer

**What's wrong**
The PR fixes drift by adding the same public-proof concern to several unrelated writer paths. That works locally, but it leaves the architecture dependent on every future writer remembering the same hidden two-graph invariant.

**Example**
A future context-graph writer now has to know that public definitions are split: discovery facts go to ONTOLOGY, while the authoritative admission proof also needs duplicated `rdf:type` and `dkg:accessPolicy` in root `_meta`. Missing any one path recreates the same drift this PR is repairing.

**Suggested direction**
Introduce a single `buildContextGraphDefinitionQuads`/`buildPublicContextGraphAuthorityQuads` boundary that owns the ONTOLOGY plus `_meta` split, then have all creation/stamping paths call it. That deletes the repeated feature-specific conditionals and makes the invariant discoverable in one place.

**For Agents**
Look at `ContextGraphMethods.createContextGraph`, the registration lazy-stamp helper, and `PublishMethods` implicit creation. Preserve the exact emitted quads, but move public-context-graph definition emission behind one canonical builder or creation service that returns both discoverability and root-authority quads. Existing tests that assert public meta proof after create/register/implicit publish should still pass.

File: packages/agent/src/context-graph-public-meta-repair.ts [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        }
+      }
+    `);
+    const existing = existingResult.type === 'bindings'
+      ? existingResult.bindings
+      : [];
+    const policies = existing
+      .filter((row) => row['predicate'] === DKG_ONTOLOGY.DKG_ACCESS_POLICY)
+      .map((row) => row['object']);
+    const hasConflictingPolicy = policies.some((value) => (
```
Comment (otReviewAgent) [inline packages/agent/src/context-graph-public-meta-repair.ts:97]: 🟡 Issue: Repair duplicates the canonical proof logic it is supposed to depend on

**What's wrong**
The repair module repeats the normalization and missing-fact rules outside `context-graph-public-meta-proof.ts`. That undercuts the stated design goal of rendering writers, validators, and repair from one model, and creates another place where public-authority semantics can drift.

**Example**
If the canonical public proof later adds another required fact or changes literal normalization, `buildAuthoritativePublicMetaAskQuery` can be updated while the repair still only understands `rdf:type` and `dkg:accessPolicy`, because the repair is not asking the proof module for a diff.

**Suggested direction**
Keep all proof matching, conflict detection, and missing-quad computation in the proof module. The repair module should only discover eligible graphs, fetch their relevant root-meta rows, and insert the missing quads returned by that canonical helper.

**For Agents**
In `context-graph-public-meta-proof.ts`, expose a small helper such as `missingAuthoritativePublicMetaQuads(contextGraphId, existingQuads)` or a reusable evaluator result that reports conflicts and missing facts. Make `repairCreatorPublicMetaProjections` feed fetched root-meta quads into that helper instead of reimplementing predicate-specific checks. Preserve current fail-closed behavior for conflicting policies.

File: packages/agent/src/context-graph-public-meta-repair.ts [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  const inserts: Quad[] = [];
+  const conflictingGraphs: string[] = [];
+  let repairedGraphs = 0;
+
+  for (const subject of candidateSubjects) {
+    if (!subject.startsWith(CONTEXT_GRAPH_URI_PREFIX)) continue;
+    const contextGraphId = subject.slice(CONTEXT_GRAPH_URI_PREFIX.length);
+    if (!contextGraphId || contextGraphDataGraphUri(contextGraphId) !== subject) continue;
+    const metaGraph = assertSafeIri(contextGraphMetaGraphUri(contextGraphId));
+    const existingResult = await store.query(`
```
Comment (otReviewAgent) [inline packages/agent/src/context-graph-public-meta-repair.ts:80]: 🟡 Issue: Batch the repair scan instead of doing sequential per-graph queries

**What's wrong**
The repair is structured as an N+1 startup loop: discover candidates once, then issue one serialized query per graph for the same predicate set. This adds avoidable orchestration complexity to boot and makes the repair harder to scale or reason about than a single batched read plus pure diff.

**Example**
With 200 creator-owned public legacy graphs, startup performs the candidate query plus 200 serialized root-meta queries, even though each per-graph query fetches the same two predicates and the results are only grouped locally before one insert.

**Suggested direction**
Use a batched projection read, optionally chunked, and keep the per-graph decision as a pure local grouping step. That makes the repair flow shorter, easier to reason about, and avoids boot-time N+1 orchestration.

**Confidence note**
This is a structure/startup maintainability concern rather than a correctness issue; the current behavior may be acceptable for small local stores.

**For Agents**
Refactor `repairCreatorPublicMetaProjections` to build eligible `(contextGraph, metaGraph)` pairs, fetch existing proof rows in batched `VALUES` queries, group by contextGraphId, then call the canonical missing/conflict helper. Preserve the single final `store.insert`/`flush` behavior and the repair result counters.

File: packages/agent/src/dkg-agent-lifecycle.ts [hunk 5]
```diff
@@ -948,6 +950,36 @@ export class LifecycleSyncMethods extends DKGAgentBase {
...
     this.log.info(ctx, `Node started, peer ID: ${this.node.peerId.toString()}`);
 
+    // Public definitions were historically written only to ONTOLOGY while
+    // late-subscriber admission requires the canonical proof in root `_meta`.
+    // Repair only graphs whose ontology creator is this exact peer. This runs
+    // after libp2p has exposed the stable peer ID but before protocol handlers
+    // and sync serving are registered. Foreign/discovered graphs remain
+    // ineligible; conflicting local policy remains fail-closed.
+    try {
+      const repaired = await repairCreatorPublicMetaProjections(
```
Comment (otReviewAgent) [inline packages/agent/src/dkg-agent-lifecycle.ts:960]: 🟡 Issue: Startup repair path is not verified end to end

**What's wrong**
This PR makes public metadata repair an automatic startup migration, but the validation only proves the helper works when called manually. That leaves the operational behavior that upgraded nodes rely on unprotected: using the live peer ID, invoking the repair at startup, and persisting the resulting `_meta` proof.

**Example**
A regression that removes or moves the line in `start()` could leave upgraded nodes serving legacy public graphs without the root `_meta` proof, while `context-graph-public-meta-repair.test.ts` would still pass because it calls the repair helper directly.

**Suggested direction**
Cover the startup hook itself, not only the helper, so the migration behavior is protected where it is actually invoked.

**For Agents**
Add a lifecycle-level regression test around `LifecycleSyncMethods.start()` or the normal test-agent startup path: seed a creator-owned public graph only in ONTOLOGY before start, use the agent's actual peer ID as creator, start the agent, then assert `buildAuthoritativePublicMetaAskQuery(contextGraphId)` returns true. Preserve the direct unit tests for repair edge cases.

**Comment by branarakic** [comment]
Closing — this PR's code is already on `main`, so merging it would change nothing.

**Evidence** (verified 2026-08-06 against `main` @ `0277d82c6`):
- the change landed on main rewritten, as `d5bec31a4`. Verified line-by-line: every non-trivial line this PR adds is present in main's copy of the touched files, with **zero missing** (`packages/agent/src/dkg-agent-context-graph.ts`, `dkg-agent-lifecycle.ts`, `dkg-agent-publish.ts`).
- `git merge-tree --write-tree origin/main 4702f2339` returns main's exact tree (`d1812f293`) — a merge would be a no-op.

Closed as part of a sweep of open PRs whose code had already landed. Please reopen if this is wrong.
